c/extras/ftrl.cc

```diff
@@ -0,0 +1,527 @@
+  PyObject* arg1;
+  PyObject* arg2;
+
+  if (!PyArg_ParseTuple(args, "OOddddLnpII:ftrl", &arg1, &arg2,
+                        &a, &b, &l1, &l2, &d, &n_epochs, &inter, &hash_type,
+                        &seed)) return nullptr;
+
+  DataTable* dt_train = py::obj(arg1).to_frame();
+  DataTable* dt_test = py::obj(arg2).to_frame();
+
+  Ftrl ft(a, b, l1, l2, d, n_epochs, inter, hash_type, seed);
+  ft.train(dt_train);
+  DataTable* dt_target = ft.test(dt_test).release();
+  py::Frame* frame_target = py::Frame::from_datatable(dt_target);
+
+  return frame_target;
+}
+
+
+/*
+*  Set up FTRL parameters and initialize weights.
+*/
+Ftrl::Ftrl(double a_in, double b_in, double l1_in, double l2_in,
+           uint64_t d_in, size_t nepochs_in, bool inter_in,
+           unsigned int hash_type_in, unsigned int seed_in) :
+  a(a_in),
+  b(b_in),
+  l1(l1_in),
+  l2(l2_in),
+  d(d_in),
+  n_epochs(nepochs_in),
+  hash_type(hash_type_in),
+  seed(seed_in),
+  inter(inter_in)
+{
+  n = DoublePtr(new double[d]);
+  w = DoublePtr(new double[d]);
+  std::memset(n.get(), 0, d * sizeof(double));
+  std::memset(w.get(), 0, d * sizeof(double));
+
+  // Initialize weights with random [0; 1] numbers
+  z = DoublePtr(new double[d]);
+  srand(seed);
+  for (uint64_t i = 0; i < d; ++i){
+    z[i] = static_cast<double>(rand()) / RAND_MAX;
+  }
+}
+
+
+/*
+*  Train FTRL model on a training dataset.
+*/
+void Ftrl::train(const DataTable* dt) {
+  // Define number of features that equal to one bias term
+  // plus `dt->ncols - 1` columns. We assume that the target column
+  // is the last one.
+  n_features = dt->ncols;
+
+  // Define number of feature interactions.
+  n_inter_features = (inter)? (n_features - 1) * (n_features - 2) / 2 : 0;
+
+  // Get the target column.
+  auto c_target = static_cast<BoolColumn*>(dt->columns[dt->ncols - 1]);
+  auto d_target = c_target->elements_r();
+
+  // Do training for `n_epochs`.
+  for (size_t i = 0; i < n_epochs; ++i) {
+    double total_loss = 0;
+    int32_t nth = config::nthreads;
+
+    #pragma omp parallel num_threads(nth)
+    {
+      // Array to store hashed features and feature interactions.
+      Uint64Ptr x = Uint64Ptr(new uint64_t[n_features + n_inter_features]);
+      // Bias term, do we need it? Results are quite similar with/without bias.
+      x[0] = 0;
+      int32_t ith = omp_get_thread_num();
+      nth = omp_get_num_threads();
+
+      for (size_t j = static_cast<size_t>(ith);
+           j < dt->nrows;
+           j+= static_cast<size_t>(nth)) {
+
+        bool target = d_target[j];
+        hash_row(x, dt, j);
+        double p = predict(x, n_features + n_inter_features);
+        double loss = logloss(p, target);
+
+        #pragma omp atomic update
+        total_loss += loss;
+
+        if ((j+1) % REPORT_FREQUENCY == 0) {
+          printf("Training epoch: %zu\tRow: %zu\tPrediction: %f\t"
+                 "Current loss: %f\tAverage loss: %f\n",
+                 i, j+1, p, loss, total_loss / (j+1));
```

Comment:
I presume this is temporary debug code?

Reply:
Yes, we will probably need some kind of progress function here. Don't know for the moment.